### PR TITLE
Expose whether the node is synced in `/v2/info`

### LINKED ISSUE #4672: [Feature request] - expose whether the node is synced in `/v2/info`
I've heard from a few different people that they'd like to have an easier way of answering the question "is my node synced with the network". One suggestion was to add this information to `/v2/info`, but I don't think that should be a requirement.

One thing I have noticed is that the prometheus metric `stacks_node_stacks_tip_height` isn't ever set until the node has finished syncing, so that does act as a kind of way to answer this question, but it would still be helpful to have other easier mechanisms.

It seems like you can answer this question by seeing if `SortitionDB::get_canonical_stacks_chain_tip_hash` returns an `Ok`, though I'm not sure.

### PR DESCRIPTION
**Pull request by ASuciuX**
### Description
Displays on the API `v2/info` whether the Stacks node is fully synced or not. I added a check based on the initial download that would specify `is_fully_synced` as true/false.

### Unexpected Resulted Flows
On an M1 Mac, I synced a Nakamoto node from 0. During this process, the block heights fluctuated unpredictably: `burn_block_height` jumped from 1000 to 4000, then to 5000, back to 2000, and so on. Eventually, it fully synced at `burn_block_height`, but `stacks_tip_height` remained at 0. The node intermittently displayed as synced and then not synced. After reaching full sync at `burn_block_height`, with `stacks_tip_height` still at 0, it continuously reported `is_fully_synced` as true.

On Linux, I synced a Nakamoto node on top of what I already had synced from a previous test. The `burn_block_height` and `stacks_tip_height` were both in progress at around 9000. The `is_fully_synced` was always true when running on this.

### Steps to reproduce
1. build from this branch
  ```   cargo build --release --bin stacks-node  ```
2. nakamoto config - copy config file and uncomment working-dir https://docs.stacks.co/nakamoto-upgrade/signing-and-stacking/sample-configuration-files#nakamoto-testnet-config

3. run node
  ```./stacks-node start --config ./node.toml    ```

4. create and run script to check
```
#!/bin/bash

while true; do
  fully=$(curl -s localhost:20443/v2/info | jq .is_fully_synced)
  btc_block=$(curl -s localhost:20443/v2/info | jq .burn_block_height)
  stx_block=$(curl -s localhost:20443/v2/info | jq .stacks_tip_height)

  if [[ "$fully" == "true" ]]; then
    echo "Fully synced at BTC" $btc_block "and STX" $stx_block
  fi

  sleep 1
done
```

### Applicable issues

- fixes #4672

### CODE REVIEW

File: stackslib/src/net/api/getinfo.rs [hunk 1]
```diff
@@ -130,6 +131,27 @@ impl RPCPeerInfoData {
...
         let public_key_hash = Hash160::from_node_public_key(&public_key);
         let stackerdb_contract_ids = network.get_local_peer().stacker_dbs.clone();
+        // let is_fully_synced = if network.work_state == PeerNetworkWorkState::BlockDownload || network.work_state == PeerNetworkWorkState::BlockInvSync {
+        //     false
+        // } else {
+        //     true
+        // };
+
+        let is_fully_synced = if let Some(block_downloader) = &network.block_downloader {
+            !block_downloader.is_initial_download()
```
Comment (ASuciuX) [inline stackslib/src/net/api/getinfo.rs:141]: I'm assuming this is not the right approach as on Linux after starting a sync on top of an existent one, it was displaying `is_fully_synced true`. Is there an easier way to test this implementation?

**Comment by wileyj** [comment]
this reminds of what pavitthra was attempting here: https://github.com/stacks-network/stacks-core/pull/2768
the solution was a little more challenging that just looking at the data you're checking for - i recall she was looking at the heights of the neighbors she was syncing with.

**Comment by hstove** [comment]
One kinda cheaty way to do this could be to utilize the `STACKS_TIP_HEIGHT_GUAGE` prometheus gauge. This is only ever set if the node is fully synced.

The naive approach would be to only support this feature with `prometheus_mod` enabled, but we could also add a static boolean that is updated in `update_stacks_tip_height`

File: stackslib/src/net/api/getinfo.rs (hunk 1 above)
Comment (jcnelson) [reply stackslib/src/net/api/getinfo.rs:141]: This is not the right approach.  You need to serve the value of `ibd` that gets passed into `PeerNetwork::run()`

**Comment by ASuciuX** [comment]
I removed my previous commits that weren't on the right track for the feature, I will open the PR with the new commit.

File: stackslib/src/net/api/getinfo.rs (hunk 1 above)
Comment (ASuciuX) [reply stackslib/src/net/api/getinfo.rs]: I've updated it by referencing the `ibd` that is used as value. I also though about saving it in PeerNetwork but that would have mean the source of truth is again split in 2 as until the `PeerNetwork::run()` is executed it has a default value. https://github.com/stacks-network/stacks-core/compare/develop...ASuciuX:stacks-core:feat/is-synced-v2-info

File: stackslib/src/net/api/callreadonly.rs [hunk 2]
```diff
@@ -194,6 +194,7 @@ impl RPCRequestHandler for RPCCallReadOnlyRequestHandler {
         preamble: HttpRequestPreamble,
         contents: HttpRequestContents,
         node: &mut StacksNodeState,
+        _ibd: bool,
```
Comment (jcnelson) [inline stackslib/src/net/api/callreadonly.rs:197]: You don't have to modify all of the calls to `try_handle_request()`.  You could instead add it as a public field to `StacksNodeState`, so interested callers can access it.

This is the purpose of the `StacksNodeState` struct -- to pass state about the node to RPC handlers in a forwards-compatible manner.

**Comment by ASuciuX** [comment]
## Results
Once Bitcoin blocks are synchronised, the IBD state is set to `false`, causing the `is_fully_synced` variable to become `true`, even though the Stacks (STX) blocks are not yet synced.
### Initial State
Initially, `is_fully_synced` was false after the bitcoin blocks were synchronised. However, after stopping the node (due to the STX blocks not syncing), `is_fully_synced` changed to true, despite the STX blocks still being unsynchronised:
```json
{
  "peer_version": 4207599114,
  "pox_consensus": "74276adcd7b47e0a4450be98646ce22b5037b95f",
  "burn_block_height": 12868,
  "stable_pox_consensus": "1973494adb7951084506f567b327816ef8a23be1",
  "stable_burn_block_height": 12867,
  "server_version": "stacks-node 0.0.1 (:, debug build, macos [aarch64])",
  "network_id": 2147483648,
  "parent_network_id": 3669344250,
  "stacks_tip_height": 0,
  "stacks_tip": "0000000000000000000000000000000000000000000000000000000000000000",
  "stacks_tip_consensus_hash": "0000000000000000000000000000000000000000",
  "genesis_chainstate_hash": "74237aa39aa50a83de11a4f53e9d3bb7d43461d1de9873f402e5453ae60bc59b",
  "unanchored_tip": null,
  "unanchored_seq": null,
  "exit_at_block_height": null,
  "is_fully_synced": true,
  "node_public_key": "027e9f66abcf6f1b49dd1fdbed860565f50047f2828b751aabaedf80254893e036",
  "node_public_key_hash": "09d51263985862fd40f384ab6c1c519228f89d27",
  "affirmations": {
    "heaviest": "nnnnnnnnnpppp",
    "stacks_tip": "",
    "sortition_tip": "nnnnnnnnnaaaaa",
    "tentative_best": "nnnnnnnnnppppa"
  },
  "last_pox_anchor": {
    "anchor_block_hash": "686f07297258943d6eaa117060ac9db1b4fdec56710c0ec7075d0940c7406461",
    "anchor_block_txid": "fdf6d6b620dc44c721419ad66de2938ad215149794bfb4853c2ef96d1cd9c4a0"
  },
  "stackerdbs": []
}
```

### Output on a New Block
1. Syncing from Genesis:
```
Fully synced at BTC 12868 and STX 0
Fully synced at BTC 12868 and STX 0
Fully synced at BTC 12868 and STX 0
Fully synced at BTC 12868 and STX 0
Fully synced at BTC 9432 and STX 0
Fully synced at BTC 10949 and STX 0
Fully synced at BTC 12529 and STX 0
Fully synced at BTC 12869 and STX 0
Fully synced at BTC 12869 and STX 0
Fully synced at BTC 12869 and STX 0
Fully synced at BTC 12869 and STX 0
```
2. Syncing from Archive:
```
Fully synced at BTC 12871 and STX 4337
Fully synced at BTC 12871 and STX 4357
Fully synced at BTC 12871 and STX 4385
Fully synced at BTC 12871 and STX 4385
Fully synced at BTC 12871 and STX 4385
Fully synced at BTC 12871 and STX 4385
Fully synced at BTC 12871 and STX 4385
Fully synced at BTC 12871 and STX 4385
```

**Comment by ASuciuX** [comment]
Hello @jcnelson,

Could you please review the changes I made regarding the retrieval of the neighbors' maximum stacks tip height? You can see the [diff here](https://github.com/stacks-network/stacks-core/compare/feat/expose-node-is-synced...feat/expose-node-is-synced-stacks).

I added an iteration through the list of neighbors to get the maximum stacks tip height. This required several modifications, as I had to adjust a trait. I think there should be an alternative way to achieve this, but I couldn't find it.

I've tested the implementation on the current testnet, but the list always returns null. I suspect the issue is with my implementation since I should at least receive data from the neighbor node that I'm syncing from, but wanted to ask if in case there could be something with the node's configurations that could create this problem. Thank you.

File: stackslib/src/net/mod.rs [hunk 3]
```diff
@@ -647,6 +647,7 @@ pub struct StacksNodeState<'a> {
     inner_mempool: Option<&'a mut MemPoolDB>,
     inner_rpc_args: Option<&'a RPCHandlerArgs<'a>>,
     relay_message: Option<StacksMessageType>,
+    ibd: bool,
```
Comment (jbencin) [inline stackslib/src/net/mod.rs:650]: IMO 3-letter variable names in struct fields should usually have a doc comment clarifying what those letters mean (especially if type is just `bool`)

```suggestion
    /// Are we in Initial Block Download (IBD) phase?
    ibd: bool,
```

**Review by jbencin — COMMENTED** [review]
Make sure to update `docs/rpc/api/core-node/get-info.schema.json` also

**Review by jbencin — APPROVED** [review]
LGTM

**Comment by ASuciuX** [comment]
> Hello @jcnelson,
> 
> 
> 
> Could you please review the changes I made regarding the retrieval of the neighbors' maximum stacks tip height? You can see the [diff here](https://github.com/stacks-network/stacks-core/compare/feat/expose-node-is-synced...feat/expose-node-is-synced-stacks).
> 
> 
> 
> I added an iteration through the list of neighbors to get the maximum stacks tip height. This required several modifications, as I had to adjust a trait. I think there should be an alternative way to achieve this, but I couldn't find it.
> 
> 
> 
> I've tested the implementation on the current testnet, but the list always returns null. I suspect the issue is with my implementation since I should at least receive data from the neighbor node that I'm syncing from, but wanted to ask if in case there could be something with the node's configurations that could create this problem. Thank you.

Hey @jbencin , I would also need an opinion on this. I have to integrate also this part ( otherwise it will only say if the bitcoin node is synced ). I made it as different PR as I'm pretty sure I'm not parsing something the best way possible but couldn't find the right spot to change. Thank you

**Review by kantai — APPROVED** [review]

**Review by jcnelson — DISMISSED** [review]

**Review by kantai — APPROVED** [review]

**Pull request merged**